### DISCUSSION TITLE
Downloading data for NP3-911-ER

### DISCUSSION THREAD

**Discussion by devamjariwala24** [ercot/api-specs#58] [Q&A] [answered]
Hi, so i am trying to download the data for NP3-911-ER but not sure how do i do it. I tried doing it with the following code with my credentials: 

import requests
from docx import Document

doc = Document()
doc.add_heading('API Output', 0)

username = 'abc@xyz.com'
password = '******'
subscription_key = 'abc123'

AUTH_URL = "https://ercotb2c.b2clogin.com/ercotb2c.onmicrosoft.com/B2C_1_PUBAPI-ROPC-FLOW/oauth2/v2.0/token"

payload = {
    'username': username,
    'password': password,
    'grant_type': 'password',
    'scope': 'openid fec253ea-0d06-4272-a5e6-b478baeecd70 offline_access',
    'client_id': 'fec253ea-0d06-4272-a5e6-b478baeecd70',
    'response_type': 'id_token'
}

auth_response = requests.post(AUTH_URL, data=payload)

if auth_response.status_code == 200:
    access_token = auth_response.json().get("id_token")
    doc.add_paragraph("Access token obtained successfully.")
else:
    doc.add_paragraph(f"Authentication failed: {auth_response.text}")

ARCHIVE_URL = "https://api.ercot.com/api/public-reports/archive"
ARCHIVE_TARGET = "NP4-732-CD"

headers = {"Authorization": "Bearer " + access_token, "Ocp-Apim-Subscription-Key": subscription_key}
product_response = requests.get(ARCHIVE_URL + "/" + ARCHIVE_TARGET, headers=headers)

doc.add_paragraph("Response from the ERCOT API:")
doc.add_paragraph(product_response.text)

doc.save("ercot_output.docx")

print("Output has been saved to 'ercot_output.docx'")

Now in the word file, I got a huge output but not sure how to navigate in it. I specifically want data of October 5, 2024 - OOctober 7, 2024. Can someone help me with it? I want the data in an excel sheet. 

This is just a sample of output which is stored in my word file: 

{"_meta":{"totalRecords":4792,"pageSize":1000,"totalPages":5,"currentPage":1,"query":{"parameterCount":0,"parameters":{},"sortedBy":"postDatetime: DESC"}},"product":{"emilId":"NP3-911-ER","name":"2-Day Ancillary Services Reports","reportTypeId":13057},"archives":[{"docId":1042628407,"friendlyName":"2_Day_AS_Disclosure","postDatetime":"2024-10-09T03:17:27.000","_links":{"endpoint":{"href":"https://api.ercot.com/api/public-reports/archive/NP3-911-ER?download=1042628407"}}},{"docId":1042360561,"friendlyName":"2_Day_AS_Disclosure","postDatetime":"2024-10-08T03:28:21.000","_links":{"endpoint":{"href":"https://api.ercot.com/api/public-reports/archive/NP3-911-ER?download=1042360561"}}},{"docId":1042083451,"friendlyName":"2_Day_AS_Disclosure","postDatetime":"2024-10-07T03:23:58.000","_links":{"endpoint":{"href":"https://api.ercot.com/api/public-reports/archive/NP3-911-ER?download=1042083451"}}} 

and continued . It is a total of 94 page document

**Comment by CaffeineLab** [comment, marked as answer]
Shameless plug ahead:  

[https://github.com/CaffeineLab/isodata
](https://github.com/CaffeineLab/isodata
)


another way to do it as follows with [playwright ](https://playwright.dev/python/docs/intro)directly:

`pip install pytest-playwright
`
```
import urllib3
from playwright.sync_api import sync_playwright


id = 'NP3-911-ER'

UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 Edg/116.0.1938.81"
url = f'https://www.ercot.com/mp/data-products/data-product-details?id={id}'

with sync_playwright() as p:
    browser = p.chromium.launch()
    page = browser.new_page(
        user_agent=UA,
        locale='en-US',
        timezone_id='America/Chicago',
    )
    page.goto(url)
    print('Wait 4s for JS to finish')
    page.wait_for_timeout(4000)
    # Get the URL of the first item in the list
    link_locators = page.locator('#reportTable ').get_by_role('link').first
    the_link = link_locators.get_attribute('href')
    browser.close()

# Grab the first file in the list
if the_link is not None:
    print(f"Download: {the_link}")
    http = urllib3.PoolManager()
    r = http.request('GET', the_link, preload_content=False)
    with open('data.zip', 'wb') as out:
        while True:
            data = r.read(65536)
            if not data:
                break
            out.write(data)
    r.release_conn()
```

**Reply by devamjariwala24** [reply]
Thank you for providing the solution. I do found a way to retrieve the data with my code.

**Comment by mackermann-ercot** [comment]
Thank you CaffeineLab!  Your are helpful as always.

Regards,
mackermann-ercot
